### PR TITLE
Permit Bootstrap deployment on Ubuntu

### LINKED ISSUE #1728: Permit the successful bootstrap on Ubuntu
**Component**: Bootstrap, package

**Why this is needed**: We want a successful bootstrap on Ubuntu, not possible today because salt formula are not adapted to Ubuntu

**What should be done**: 
Modify Salt formulas to permit Ubuntu bootstrap

**Implementation proposal** (strongly recommended):
- Modify salt-formulas

**Test plan**: Launch the bootstrap script on an Ubuntu node and verify Kube and all other packages are well installed.

### PR DESCRIPTION
**Pull request by ycointe**
**Component**: Salt

**Context**: MetalK8s on Ubuntu

**Summary**: We want to deploy the Bootstrap node on Ubuntu

**Acceptance criteria**: Launch the bootstrap.sh script and deploy bootstrap node on Ubuntu

---

Closes: #1728

### CODE REVIEW

File: salt/_modules/metalk8s_package_manager_apt.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
+'''
+Describes our custom way to deal with yum packages
+so that we can support downgrade in metalk8s
+'''
+import logging
+# pylint: disable=import-error
+import apt      # type: ignore
+import apt_pkg  # type: ignore
+# pylint: enable=import-error
```
Comment (TeddyAndrieux) [inline salt/_modules/metalk8s_package_manager_apt.py]: Please do like for `salt/_modules/metalk8s_cordon.py` using `HAS_LIBS` because here if you not have apt you will get errors

File: salt/_modules/metalk8s_package_manager_apt.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+#                pkg_name,
+#                ret['stderr'] or ret['stdout']
+#            )
+#            return None
+
+ #       is_installed = bool(ret['stdout'].strip())
+#        if not is_installed and pkg_name != name:
+            # Any package requiring the target `name` that is not yet installed
+            # should not be installed
+#            del all_pkgs[pkg_name]
```
Comment (alexandre-allard) [inline salt/_modules/metalk8s_package_manager_apt.py]: Please remove all the commented code

File: salt/_modules/metalk8s_package_manager_apt.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+#    dependents = _list_dependents(
+#        name,
+#        version,
+#        fromrepo=fromrepo,
+#        allowed_versions=versions_dict,
+#    )
+#
+#    all_pkgs.update(dependents)
+
```
Comment (alexandre-allard) [inline salt/_modules/metalk8s_package_manager_apt.py]: Remove this, it'll be handled in another PR I think

File: salt/_modules/metalk8s_package_manager_yum.py [hunk 4]
```diff
@@ -11,8 +11,9 @@
 
 
 def __virtual__():
-    return __virtualname__
-
+    if __grains__['os_family'].lower() == 'redhat':
```
Comment (alexandre-allard) [inline salt/_modules/metalk8s_package_manager_yum.py]: You don't need lower() as we already know how it's written:
```
if __grains__['os_family'] == 'RedHat':
```

File: salt/metalk8s/map.jinja [hunk 5]
```diff
@@ -2,13 +2,31 @@
...
   },
-} %}
+}, grain='os', merge=defaults.get('version_defaults')) %}
+
+{% set packages = salt['grains.filter_by']({
+  'CentOS': {},
+  'Ubuntu': {
+    'm2crypto': 'python-m2crypto',
+    'iproute': 'iproute2',
+
```
Comment (alexandre-allard) [inline salt/metalk8s/map.jinja]: extra line here

File: salt/metalk8s/map.jinja [hunk 6]
```diff
@@ -24,23 +42,139 @@
...
+        'enabled': 0,
+      },
     }
   },
-}, merge=defaults.get('repo')) %}
+  'Ubuntu': {
+    'repositories': {
+      'metalk8s-bionic': {
+        'name': 'deb [trusted=yes] '
+              'file:///vagrant/_build/root/packages/debian/metalk8s-bionic '
```
Comment (alexandre-allard) [inline salt/metalk8s/map.jinja]: You should not store the hardcoded path here, it must be computed in the salt formula instead.

File: salt/metalk8s/map.jinja [hunk 7]
```diff
@@ -24,23 +42,139 @@
...
+        'repo_gpg_check': 0,
+        'enabled': 0,
+      },
     }
   },
-}, merge=defaults.get('repo')) %}
+  'Ubuntu': {
+    'repositories': {
+      'metalk8s-bionic': {
+        'name': 'deb [trusted=yes] '
```
Comment (alexandre-allard) [inline salt/metalk8s/map.jinja]: Should be a dict of options (and then rendered in the formula):
```
  'Ubuntu': {
    'repositories': {
      'metalk8s-bionic': {
        'options': {
          'trusted': 'yes',
        },
```

File: salt/metalk8s/repo/ubuntu.sls [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  file.managed:
+    - name: /etc/apt/preferences.d/1001-metalk8s
+    - contents: |
+        Package: *
+        Pin: release l=metalk8s
+        Pin-Priority: 1001
+
+Repositories configured:
+  test.succeed_without_changes:
+    - require:
```
Comment (alexandre-allard) [inline salt/metalk8s/repo/ubuntu.sls]: Must also require the state above `- file: Add APT metalk8s repositories preferences`

File: salt/metalk8s/repo/ubuntu.sls [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+{%- set repo_host = pillar.metalk8s.endpoints['repositories'].ip %}
+{%- set repo_port = pillar.metalk8s.endpoints['repositories'].ports.http %}
+
+Install python-apt:
+  pkg.installed:
+    - name: python-apt
+
+{%- for repo_name, repo_config in repo.repositories.items() %}
+Configure {{repo_name}} repository:
```
Comment (alexandre-allard) [inline salt/metalk8s/repo/ubuntu.sls]: Please, for consistency, put spaces between the jinja markups as it's done everywhere else.

File: salt/metalk8s/repo/ubuntu.sls [hunk 10]
```diff
@@ -0,0 +1,33 @@
+{%- from "metalk8s/map.jinja" import metalk8s with context %}
+{%- from "metalk8s/map.jinja" import repo with context %}
+
+{%- set repo_host = pillar.metalk8s.endpoints['repositories'].ip %}
+{%- set repo_port = pillar.metalk8s.endpoints['repositories'].ports.http %}
```
Comment (alexandre-allard) [inline salt/metalk8s/repo/ubuntu.sls]: These variables are not used anywhere

File: salt/metalk8s/repo/ubuntu.sls [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+Install python-apt:
+  pkg.installed:
+    - name: python-apt
+
+{%- for repo_name, repo_config in repo.repositories.items() %}
+Configure {{repo_name}} repository:
+  pkgrepo.managed:
+    - name : {{repo_config.name}}
```
Comment (alexandre-allard) [inline salt/metalk8s/repo/ubuntu.sls]: So you only handle local repositories, it seems that deploying on bootstrap node was the scope of the linked issue, but hey it does not cost much to handle http served repositories.

File: salt/metalk8s/map.jinja [hunk 12]
```diff
@@ -24,23 +41,139 @@
...
   },
-}, merge=defaults.get('repo')) %}
+  'Ubuntu': {
+    'repositories': {
+      'metalk8s-bionic': {
+        'options': {
+          'trusted': 'yes',
+        },
+        'comps': [
+            'metalk8s-bionic'
```
Comment (alexandre-allard) [inline salt/metalk8s/map.jinja]: Too much indentation

File: salt/metalk8s/map.jinja [hunk 13]
```diff
@@ -24,23 +41,139 @@
...
+        },
+        'comps': [
+            'metalk8s-bionic-backports'
+        ]
+      },
+      'metalk8s-salt_ubuntu1804': {
+        'options': {
+          'trusted': 'yes',
+        },
+         'comps': [
```
Comment (alexandre-allard) [inline salt/metalk8s/map.jinja]: extra space before `'comps'`

File: salt/_modules/metalk8s_package_manager_apt.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
+'''
+Describes our custom way to deal with yum packages
+so that we can support downgrade in metalk8s
+'''
+import logging
+HAS_LIBS = True
+import apt
+import apt_pkg
+HAS_LIBS = False
```
Comment (alexandre-allard) [inline salt/_modules/metalk8s_package_manager_apt.py]: No, no, no, no, no !
```
HAS_APT_LIBS = True
try:
    import apt
    import apt_pkg
except ImportError:
    HAS_APT_LIBS = False
```
Btw, `apt_pkg` does not seem to be used anywhere.

File: salt/_modules/metalk8s_package_manager_apt.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+except ImportError:
+    HAS_APT_LIBS = False
+
+log = logging.getLogger(__name__)
+
+__virtualname__ = 'metalk8s_package_manager'
+
+
+def __virtual__():
+    if __grains__['os_family'].lower() == 'debian':
```
Comment (TeddyAndrieux) [inline salt/_modules/metalk8s_package_manager_apt.py]: It's better to me if you use `HAS_APT_LIBS` here

File: salt/metalk8s/map.jinja [hunk 16]
```diff
@@ -2,13 +2,30 @@
...
+  },
+  'Ubuntu': {
+    'kubernetes': versions.kubernetes,
+    'repo': {
+      'packages': versions.packages.ubuntu,
+      'images': versions.images,
+    },
   },
-} %}
+}, grain='os', merge=defaults.get('version_defaults')) %}
```
Comment (TeddyAndrieux) [inline salt/metalk8s/map.jinja]: I suggest to use `os_family` instead of `os` as we expect that most of the code will work for RedHat and CentOS

File: salt/metalk8s/map.jinja [hunk 17]
```diff
@@ -24,23 +41,145 @@
 }, merge=defaults.get('metalk8s')) %}
 
 {% set kubernetes = salt['grains.filter_by']({
-  'default': {}
-}, merge=defaults.get('kubernetes')) %}
+  'Ubuntu': {}
+}, grain='os', merge=defaults.get('kubernetes')) %}
```
Comment (TeddyAndrieux) [inline salt/metalk8s/map.jinja]: Why this change ? We want to merge this variable also for RedHat. If you have nothing special to override here keep the old one please :)

File: salt/metalk8s/map.jinja (hunk 16 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/map.jinja]: PS: If you give no `grain` arg it's `os_family` by default https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.grains.html#salt.modules.grains.filter_by

File: salt/metalk8s/repo/ubuntu.sls [hunk 18]
```diff
@@ -0,0 +1,59 @@
+{%- from "metalk8s/map.jinja" import metalk8s with context %}
+{%- from "metalk8s/map.jinja" import repo with context %}
+
+{%- set repo_host = pillar.metalk8s.endpoints['repositories'].ip %}
+{%- set repo_port = pillar.metalk8s.endpoints['repositories'].ports.http %}
+
+Install python-apt:
+  pkg.installed:
+    - name: python-apt
```
Comment (TeddyAndrieux) [inline salt/metalk8s/repo/ubuntu.sls]: Here you expect that some repo available before our configuration with `python-apt` available. Please remove this and just assume that `python-apt` is installed for the moment (it's installed with salt, so for the bootstrap node it's ok)

File: salt/metalk8s/repo/ubuntu.sls [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+        Pin: release l=metalk8s
+        Pin-Priority: 1001
+
+Repositories configured:
+  test.succeed_without_changes:
+    - require:
+      - file: Add APT metalk8s repositories preferences
+{%- for repository_name in repo.repositories.keys() %}
+      - file: Configure {{ repository_name }} repository
+{%- endfor %}
```
Comment (TeddyAndrieux) [inline salt/metalk8s/repo/ubuntu.sls]: For me should be better to use `require_in` instead of doing two time the `for` loop here (but I know it's the same for RedHat for the moment :stuck_out_tongue: )

File: salt/metalk8s/repo/centos.sls [hunk 20]
```diff
@@ -13,12 +13,12 @@ Install yum-plugin-versionlock:
 
 {%- for repo_name, repo_config in repo.repositories.items() %}
   {%- if repo.local_mode %}
-    {%- set repo_base_url = "file://" ~ 
+    {%- set repo_base_url = "file://" ~
                             salt.metalk8s.get_archives()[saltenv].path ~ "/" ~
-                            repo.relative_path %}
+                            repo.relative_path ~ "/" ~ ['os_family'].lower() %}
```
Comment (alexandre-allard) [inline salt/metalk8s/repo/centos.sls]: ```grains['os_family'].lower()```

File: salt/metalk8s/map.jinja [hunk 21]
```diff
@@ -55,8 +194,21 @@
...
+    }
+  },
+  'CentOS': {
+    'service': {
+      'options': {
+        'resolv-conf': '/etc/resolv.conf'
+      }
+    }
+  }
+}, grain='os', merge=defaults.get('kubelet')) %}
```
Comment (TeddyAndrieux) [inline salt/metalk8s/map.jinja]: Please use `osfamily` (or just remove `grain='os'` as `osfamily` is the default value) everywhere or @alexandre-allard-scality  change for redhat will be broken in 2.4

Even if you (or Alexandre) are going to have conflicts in any case I think.

File: salt/metalk8s/repo/init.sls [hunk 22]
```diff
@@ -1,2 +1,2 @@
 include:
-  - .offline
+  - .{{ grains['os'] | lower }}
```
Comment (TeddyAndrieux) [inline salt/metalk8s/repo/init.sls]: For me you should use `osfamily` here also

**Review by cmonfort — APPROVED** [review]
